crates/agglayer-node/src/rpc/error.rs

```diff
@@ -0,0 +1,88 @@
+//! Support for structured errors in RPC.
+
+use jsonrpsee::types::error::ErrorObjectOwned;
+use serde::Serialize;
+use serde_json::value::RawValue;
+
+/// RPC error codes
+pub mod code {
+    use crate::kernel::SettlementError;
+
+    // Rollup auth errors
+    pub const ROLLUP_NOT_REGISTERED: i32 = -10101;
+    pub const ROLLUP_SIGNATURE: i32 = -10102;
+
+    // Verification errors
+    pub const PROOF_ETH: i32 = -10201;
+    pub const PROOF_ZKEVM: i32 = -10202;
+
+    // Settlement errors
+    pub const NO_RECEIPT: i32 = -10301;
+    pub const PROVIDER_ERROR: i32 = -10302;
+    pub const CONTRACT_ERROR: i32 = -10303;
+
+    // Status retrieval errors
+    pub const STATUS_GET_TX: i32 = -10401;
+    pub const STATUS_GET_BLOCK: i32 = -10402;
+    pub const STATUS_TX_NOT_FOUND: i32 = -10403;
+
+    /// Error code for given settlement error
+    pub fn settlement<Rpc: ethers::middleware::Middleware>(err: &SettlementError<Rpc>) -> i32 {
+        match err {
+            SettlementError::NoReceipt => NO_RECEIPT,
+            SettlementError::ProviderError(_) => PROVIDER_ERROR,
+            SettlementError::ContractError(_) => CONTRACT_ERROR,
+        }
+    }
+}
+
+/// Represents extra structured data that may be attached to a jsonrpc error.
+#[derive(Clone, Debug)]
+pub struct ErrorData(Option<Box<RawValue>>);
+
+impl ErrorData {
+    /// No error data.
+    pub const EMPTY: Self = Self(None);
+
+    /// New structured error data.
+    pub fn new<T: ?Sized + Serialize>(data: &T) -> Self {
+        Self(serde_json::value::to_raw_value(data).ok())
+    }
+
+    fn into_raw_value(self) -> Option<Box<RawValue>> {
+        self.0
+    }
+}
+
+/// A top-level RPC error.
+#[derive(thiserror::Error, Debug)]
+#[error(transparent)]
+pub struct RpcError(ErrorObjectOwned);
+
+impl RpcError {
+    /// Create an [RpcError] from a jsonrpsee error object.
+    pub fn form_error_object(err_obj: ErrorObjectOwned) -> Self {
+        Self(err_obj)
+    }
+
+    /// Create a basic [RpcError] with just an error code and a message.
+    pub fn basic(code: i32, message: impl ToString) -> Self {
```

Comment:
```suggestion
    pub fn basic<C: Into<i32>>(code: C, message: impl ToString) -> Self {
```

Reply:
I think it's better to keep the interfaces tight, only allowing things that make sense. There is, for example, a `From<bool>` for `i32`. The suggested change would allow `RpcError::basic(false, "foo")`. It's not clear at all that the first argument is some sort of error code.

I'm actually thinking of going in the opposite direction: Introduce a new type for agglayer-specific error codes and accept it here rather than plain `i32`, so that using the named constants is encouraged.

Reply:
Yes, we can introduce an enum `RpcCode` that resolve to the constant and either impl `From` to use it in the various needed method or have a custom method for it:

```rust
pub const ROLLUP_NOT_REGISTERED: i32 = -10101;

pub enum RpcCode {
  RollupNotRegistered
}

impl From<RpcCode> for i32 {
  fn from(value: RpcCode) -> i32 {
    match value {
      RpcCode::RollupNotRegistered => ROLLUP_NOT_REGISTERED
    }
  }
}

```

Reply:
So I did something like this but with a struct which is just a wrapper around the code, with a bunch of associated constants. The struct is a bit more open-ended in that it still allows being constructed form arbitrary integers. Over all it probably does not make a massive amount of difference.

Reply:
Resolving this thread as the discussion does not apply anymore